### PR TITLE
Add 2 new rules NestedFunctionNames and UnnestedFunctionNames

### PR DESCRIPTION
**Pull request by su8898**
These rules allow configuring naming conventions for nested and unnested function names.

### CODE REVIEW

File: docs/content/how-tos/rules/FL0075.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+---
+title: FL0075
+category: how-to
+hide_menu: true
+---
+
+# UnnestedFunctionNames (FL0075)
+
+*Introduced in `0.20.3`*
```
Comment (siwatanejo) [inline docs/content/how-tos/rules/FL0075.md:9]: please change this to 0.21.1

File: docs/content/how-tos/rules/FL0076.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
+---
+title: FL0076
+category: how-to
+hide_menu: true
+---
+
+# NestedFunctionNames (FL0076)
+
+*Introduced in `0.20.3`*
```
Comment (siwatanejo) [inline docs/content/how-tos/rules/FL0076.md:9]: same here

File: docs/content/how-tos/rules/FL0075.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+## Rule Settings
+
+    {
+        "unnestedFunctionNames": {
+        "enabled": false,
+        "config": {
+            "naming": "PascalCase",
+            "underscores": "None"
+        }
+    }
```
Comment (knocte) [inline docs/content/how-tos/rules/FL0075.md:32]: please remove trailing whitespace from the line above

File: docs/content/how-tos/rules/FL0076.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+## Rule Settings
+
+    {
+        "NestedFunctionNames": {
+        "enabled": false,
+        "config": {
+            "naming": "CamelCase",
+            "underscores": "None"
+        }
+    }
```
Comment (knocte) [inline docs/content/how-tos/rules/FL0076.md:32]: same here

File: src/FSharpLint.Core/Rules/Conventions/Naming/NamingHelper.fs [hunk 5]
```diff
@@ -361,3 +361,14 @@ let rec identFromSimplePat = function
...
+
+let rec isNested args nodeIndex =
+    let parent = args.SyntaxArray.[nodeIndex].ParentIndex
+    let actual = args.SyntaxArray.[parent].Actual
+
+    match actual with
+    | AstNode.ModuleDeclaration (SynModuleDecl.Types (_, _)) -> false
+    | AstNode.ModuleDeclaration (SynModuleDecl.NestedModule (_, _, _, _, _)) -> false
+    | AstNode.Binding (SynBinding (_, _, _, _, _, _, _, _, _, _, _, _)) -> false
+    | AstNode.Expression (SynExpr.LetOrUse (_, _, _, _, _)) -> true
```
Comment (knocte) [inline src/FSharpLint.Core/Rules/Conventions/Naming/NamingHelper.fs:373]: this match block can be hugely simplified if you move the case `AstNode.Expression (SynExpr.LetOrUse (_, _, _, _, _))` to the first position, and then just use `_ -> false` as a second case

File: src/FSharpLint.Core/Rules/Conventions/Naming/UnnestedFunctionNames.fs [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+            match List.tryLast longIdent.Lid with
+            | Some ident -> (ident, ident.idText, None) |> Array.singleton
+            | None -> Array.empty
+        | _ -> Array.empty
+    | _ -> Array.empty
+
+let private getIdentifiers (args: AstNodeRuleParams) =
+    match args.AstNode with
+    | AstNode.Binding (SynBinding (_, _, _, _, _attributes, _, _, pattern, _, _, _, _)) ->
+        match isNested args args.NodeIndex with
```
Comment (knocte) [inline src/FSharpLint.Core/Rules/Conventions/Naming/UnnestedFunctionNames.fs:23]: please don't use match blocks for true/false values, just use an `if ... else`

**Comment by su8898** [comment]
@knocte thanks for the review. I've updated the PR with your suggestions. Could you please take a look?

File: src/FSharpLint.Core/Rules/Conventions/Naming/NestedFunctionNames.fs [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+            match List.tryLast longIdent.Lid with
+            | Some ident -> (ident, ident.idText, None) |> Array.singleton
+            | None -> Array.empty
+        | _ -> Array.empty
+    | _ -> Array.empty
+
+let private getIdentifiers (args: AstNodeRuleParams) =
+    match args.AstNode with
+    | AstNode.Binding (SynBinding (_, _, _, _, _attributes, _, _, pattern, _, _, _, _)) ->
+        match isNested args args.NodeIndex with
```
Comment (knocte) [inline src/FSharpLint.Core/Rules/Conventions/Naming/NestedFunctionNames.fs:23]: this is another match block on true/false, please convert into if else

Comment (su8898) [reply src/FSharpLint.Core/Rules/Conventions/Naming/NestedFunctionNames.fs]: I've updated the PR with this fix. Could you please take a look? thanks

File: src/FSharpLint.Core/Rules/Conventions/Naming/UnnestedFunctionNames.fs [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+let private getFunctionIdents _ =
+    function
+    | SynPat.LongIdent (longIdent, _, _, pats, _, _) ->
+        match pats with
+        | SynArgPats.Pats _ ->
+            match List.tryLast longIdent.Lid with
+            | Some ident -> (ident, ident.idText, None) |> Array.singleton
+            | None -> Array.empty
+        | _ -> Array.empty
+    | _ -> Array.empty
```
Comment (knocte) [inline src/FSharpLint.Core/Rules/Conventions/Naming/UnnestedFunctionNames.fs:18]: this function is exactly the same in UnnestedFunctionNames.fs and NestedFunctionNames.fs, let's place it in a common place to share it between the two

Comment (su8898) [reply src/FSharpLint.Core/Rules/Conventions/Naming/UnnestedFunctionNames.fs]: I've placed the function in a common place and updated the PR. Could you please take a look? thanks

File: src/FSharpLint.Core/Rules/Conventions/Naming/NamingHelper.fs [hunk 9]
```diff
@@ -361,3 +361,22 @@ let rec identFromSimplePat = function
...
+let rec isNested args nodeIndex =
+    let parent = args.SyntaxArray.[nodeIndex].ParentIndex
+    let actual = args.SyntaxArray.[parent].Actual
+
+    match actual with
+    | AstNode.Expression (SynExpr.LetOrUse (_, _, _, _, _)) -> true
+    | _ -> false
+
+let getFunctionIdents _ =
+    function
```
Comment (knocte) [inline src/FSharpLint.Core/Rules/Conventions/Naming/NamingHelper.fs:374]: I'm not a fan of using the keyword `function` here, can you just write a normal function?

File: src/FSharpLint.Core/Rules/Conventions/Naming/UnnestedFunctionNames.fs [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+open FSharpLint.Framework.Rules
+open FSharpLint.Rules.Helper.Naming
+
+let private getIdentifiers (args: AstNodeRuleParams) =
+    match args.AstNode with
+    | AstNode.Binding (SynBinding (_, _, _, _, _attributes, _, _, pattern, _, _, _, _)) ->
+        if isNested args args.NodeIndex then
+            Array.empty
+        else
+            getPatternIdents Accessibility.Public getFunctionIdents true pattern
```
Comment (knocte) [inline src/FSharpLint.Core/Rules/Conventions/Naming/UnnestedFunctionNames.fs:15]: are we sure we want to use Accessibility.Public here? we want all functions, even private and internal ones

Comment (su8898) [reply src/FSharpLint.Core/Rules/Conventions/Naming/UnnestedFunctionNames.fs:15]: I don't think the `Accessibility.Public` is used in this case. You think it's best to add a `Both` or `NotApplicable` member to the `Accessibility` DU?

Comment (knocte) [reply src/FSharpLint.Core/Rules/Conventions/Naming/UnnestedFunctionNames.fs:15]: I don't understand how can this be not used?

Comment (webwarrior-ws) [reply src/FSharpLint.Core/Rules/Conventions/Naming/UnnestedFunctionNames.fs:15]: Passing `Accessibility.Public` here means we start within public scope.
So it's `getFunctionIdents` function that can decide what is included based on identifier's accessibility.
Which ignores accessibility parameter, and processes all functions.

File: src/FSharpLint.Core/Rules/Conventions/Naming/NamingHelper.fs (hunk 9 above)
Comment (webwarrior-ws) [reply src/FSharpLint.Core/Rules/Conventions/Naming/NamingHelper.fs:374]: Yes, and better to rename `_` to `accessibility` or `_accessibility` just to see what is being ignored

**Pull request closed without merging**